### DISCUSSION TITLE
Requirements

### DISCUSSION THREAD

**Discussion by fredchristianson** [fredchristianson/javascript-game-engine#4] [General]
There are 4 parts to the first version
1. A simple HTTP server for static files and API
2. Game Management
3. Game Core
4. Game(s)

## File Server
The file server must be as easy as possible to run on any collaborator's machine.  It does not have to be part of the repository.
It must serve these types of files at a minimum
- .html
- .css
- .js
- .json
- images (jpg, png, gif)

Ideally it can server fonts, audio and any other media.

It should implement an API to return a list of games.

## Game Management
This part of the system will manage
- players
- game statics (# times played, win/loss counts, high scores)
- user preferences
- game settings (high/low res, speed, difficulty, etc)
- game control (start, stop, pause)
- game state (saving state to resume later)

## Game Core
These are the objects and functions that are used to build games.
- World
- Camera
- Component
- Actor
- Mechanics
- Physics

**World**
A world consists of one or more layers.  
Layers can contain
- background image
- effects (drawn with javascript)
- Components (cards, dice, pegs, etc)
- Actors (human and game)

**Camera** 
A camera is a view of the world projected onto the browser window.  There will be multiple types of camera functionality
- simple (one location)
- zoomable
- movable (varying location)
- perspective (world appears to have depth)

All cameras will have "simple" functionality.  One or more of the other functions can be implemented for a game.
A game may have multiple cameras

**Component**
A game component is any piece of a game.  Some examples are 
- dice
- cards
- tiles
All components have a 2D view.  A component may provide a 3D view.
Views may be 
- static image
- a set of images (multiple files or one file with rows and/or columns)
- drawn with javascript (HTML, CSS, canvas, SVG)
Components can be animated by time, location, or custom javascript.

**Actors**
These are components with actions.  Actions may be controlled by human input (mouse, keyboard, etc).  Or may be controlled by javascript.  Javascript actors may be 
- scripted
- random
- custom

**Mechanics**
Mechanics functionality controls gameplay.  The primary mechanics are
- turn-based
- realtime
The core mechanics will also manage rules including
- win/lose
- timers
- TBD

**Physics**
Movement of components is managed by the physics module.  It will at least handle
- Controlled movement (human or javascript actions)
- Automated movement (non-actor components)
- Collisions
In addition it may handle modifications to movement
- changes based on environment (wind, ice, etc)
- changes based on actor state (damages, hungry, etc)

## Games
A game implementation provides one or more core objects
- World
- Camera
- Component
- Actor
- Mechanics
- Physics
The game core will isolate game javascript and DOM as much as possible.

# Future
In versions after 1, we may want to add
- multiplayer games
    - Through server
    - With WebRTC
- Camera/Microphone
- External datasources (gamify excel data?)
- External APIs (set game weather by local weather)
When V1 decisions may impact the effort to implement future requirements, it's better to make these features easier.

**Comment by chylinski82** [comment]
- When you say that the http server doesn't not need to be part of the repository, does it mean that we won't work on a server that all of us will use to serve files, but that each of us can or should have a custom solution?

- Your blog is about the project being as modular as possible to avoid unnecessary complications, to the point that individual objects/classes should have as few methods as possible and, if i understood it right, that's where the 'public virtual methods' come as an interface between different objects. Would that mean, that for example an actor wouldn't have a method 'kick' in order to interact with game component 'ball', but would use and independent function? like a function from 'physics' module?

**Reply by fredchristianson** [reply]
I created a minimal nodejs server.  For now it just server static files in the browser folder.  Everyone can use that.  We may want to improve it.
My initial goal was to load files from "file://" URLs.  But it requires to many disabling of browser security flags, so a minimal nodejs server seems best.

**Reply by chylinski82** [reply]
sure

**Comment by fredchristianson** [comment]
My first thought in your example is to make a KickAction class.  A game can create a SportActor object and add KickAction to it.  Depending on the game, KickAction may be guaranteed contact, or it would move the leg/foot and the physics module would determine if there is contact with a ball and what happens next (in coordination with the game, or with physics actions created by the game).

My process is to break the act of kicking into the appropriate number of classes.  There is a lot of debate over what is appropriate.  Some people argue that "fat classes" are more powerful - A SportActor should be able to kick and throw and run and swim even though not all SportActors do that.  I prefer many little classes that work together.

**Reply by chylinski82** [reply]
I'm trying to get my head around it, so kicking wouldn't be a method but a class? how would that work? SportActor comes within a short distance of BallComponent and as the controls send information to KickAction and KickAction communicates with physics module (which in this case acts as interface between classes?) and the physics module is the one that actualy dispatches the real action?
